File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/others/distance/NearestByDepot.java
package cujae.inf.ic.om.heuristic.assignment.others.distance;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;

import cujae.inf.ic.om.heuristic.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar los clientes a los depósitos en el orden de la lista de depósitos partiendo del criterio de cercanía a los clientes*/
public class NearestByDepot extends ByNotUrgency {
	
	private Solution solution = new Solution();	

	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	private ArrayList<Integer> listIDDepots;
	private NumericMatrix costMatrix;
	
	private RowCol rcBestCustomer = null;  
	private int idDepot = -1;
	private int posDepot = 0;
	private double capacityDepot = 0.0;	
	private int posDepotMatrix = -1;
	private int idCustomer = -1;
	private int posCustomer = -1;
	private double requestCustomer = 0.0;
	
	public int posCluster = -1;
	public double requestCluster = 0.0;
	
	private boolean isNext = true;
	
	public NearestByDepot() {
		super();
	}
	
	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}	
	
	@Override
	public void initialize() {
		listClusters = initializeClusters();
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());
		costMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
	}
	
	@Override
	public void assign() {		
		
		int totalCustomers = listCustomersToAssign.size();
		int totalDepots = listIDDepots.size();
		
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()) && (!costMatrix.fullMatrix(Double.POSITIVE_INFINITY)))
		{	
			if(isNext)
			{
				idDepot = listIDDepots.get(posDepot);
				capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idDepot));
				posDepotMatrix = Problem.getProblem().getPosElement(idDepot);
			}
			
			rcBestCustomer = costMatrix.indexLowerValue(posDepotMatrix, 0, posDepotMatrix, (totalCustomers - 1));

			posCustomer = rcBestCustomer.getCol();
			idCustomer = Problem.getProblem().getCustomers().get(posCustomer).getIDCustomer();
			requestCustomer = Problem.getProblem().getCustomers().get(posCustomer).getRequestCustomer();

			costMatrix.setItem(posDepotMatrix, posCustomer, Double.POSITIVE_INFINITY);
			
			int posCluster = findCluster(idDepot, listClusters);

			if(posCluster != -1) 
			{
				double requestCluster = listClusters.get(posCluster).getRequestCluster();
				
				if(capacityDepot >= (requestCluster + requestCustomer)) 
				{
					requestCluster += requestCustomer;

					listClusters.get(posCluster).setRequestCluster(requestCluster);
					listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

					costMatrix.fillValue(totalCustomers, posCustomer, (totalCustomers + totalDepots - 1), posCustomer, Double.POSITIVE_INFINITY);

					listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));

					if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
					{
						if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(listClusters.remove(posCluster));
						else
							listClusters.remove(posCluster);
						
						listIDDepots.remove(posDepot);
						posDepot--;
					}
					
					posDepot++;

					if(posDepot >= listIDDepots.size())
						posDepot = 0;
					
					isNext = true;
				}	
				else
				{
					isNext = false;
					
					if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
					{
						if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(listClusters.remove(posCluster));
						else
							listClusters.remove(posCluster);
						
						listIDDepots.remove(posDepot);
						posDepot--;
						
						posDepot++;

						if(posDepot >= listIDDepots.size())
							posDepot = 0;
						
						isNext = true;
					}
				}
			}
		}	
	}	
	
	@Override
	public Solution finish() {
		
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!listClusters.get(k).getItemsOfCluster().isEmpty())
					solution.getClusters().add(listClusters.get(k));

		return solution;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/others/distance/nearest_by_depot.py
import numpy as np
from typing import List
from ...classical.by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.solution.solution import Solution
from ....problem.solution.cluster import Cluster

class NearestByDepot(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_id_depots: List[int]
    
    def to_clustering(self):
        self.initialize()
        self.assign()
        return self.finish()
        
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_id_depots = list(Problem.get_problem().get_list_id_depots())
        
    def assign(self):
        cost_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        current_depot_index = 0
        
        while any(self.list_customers_to_assign) and self.list_clusters:
            id_depot = self.list_id_depots[current_depot_index]
            current_depot = Problem.get_problem().get_depot_by_id_depot(id_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(current_depot)
  
            cost_row = cost_matrix[current_depot_index, :]
            col = np.argmin(cost_row)
            
            selected_customer = self.list_customers_to_assign[col]    
            request_customer = selected_customer.get_request_customer()

            pos_cluster = self.find_cluster(id_depot, self.list_clusters)
                
            if pos_cluster != -1:
                cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()

                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(selected_customer.get_id_customer())

                    capacity_depot -= request_customer
                    self.list_customers_to_assign[col] = None

                    if self.is_full_depot(self.list_customers_to_assign, cluster.get_request_cluster(), capacity_depot):
                        if cluster.get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                    
                    cost_matrix[:, col] = float('inf')
                    
            current_depot_index = (current_depot_index + 1) % len(self.list_id_depots)
                    
    def finish(self) -> Solution:
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
        
        OSRMService.clear_distance_cache()
        
        return self.solution